File: python3/minimum_deletions_to_make_character_frequencies_unique.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def minDeletions(self, s: str) -> int:
        ans = 0
        d = {}

        for l in s:
            if l not in d:
                d[l] = 1
            else:
                d[l] += 1

        mem = set()
        for v in d.values():
            while v > 0 and v in mem:
                v -= 1
                ans += 1
            mem.add(v)

        return ans

File: golang/minimum_deletions_to_make_character_frequencies_unique.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func minDeletions(s string) int {
	ans := 0

	charMap := map[rune]int{}
	for _, c := range s {
		charMap[c]++
	}

	memo := map[int]int{}
	for _, val := range charMap {
		for val > 0 {
			if _, ok := memo[val]; ok {
				val--
				ans++
			} else {
				break
			}
		}

		memo[val] = 1
	}

	return ans
}
